2015/presentations/condemned_to_reinvent_sql/text.md

```diff
@@ -0,0 +1,342 @@
+# Condemned to re-invent SQL, poorly - Michał Lowas-Rzechonek
+
+## Introduction
+
+The days when developers were expected to write SQL by hand are long gone. ORMs
+are getting more and more popular, especially in web development.
+
+While it's hard to beat the convenience of using an ORM, they lure us into
+thinking about a database as a simple "object storage". One of the problems is
+the fact that in most ORMs, querying the database happens pretty much
+exclusively via a model class, which constrains the set of results to known
+attributes of our objects.
+
+This might be one of the reasons why NoSQL approaches are more alluring: just
+throw your objects into a bin and save them on the disk. Indeed, there is no
+need to worry about tables, constraints and indexes if all you do is load data
+from the storage and process it inside the app!
+
+I think the "object storage" way of thinking servely limits our ability to
+process the data. In a relational model, columns are computed ad-hoc, when
+returning query result, not when defining the schema. This is one of the
+aspects of so-called "object-relational impedance mismatch" that results in
+much headache for people smarter than us.
+
+In this article, I'd like to show a few features of relational model which,
+while mapping poorly to object-oriented world, give the programmer very
+powerful data manipulation tools.
+
+I'm also going to suggest ways to map these features into application code.
+While I believe the complete, robust translation is not achievable for reasons
+mentioned above, there are techniques that allow us to tie the data and
+application layers together, especially in a flexible, multi-paradigm language
+like Python.
+
+## Views
+
+Consider an university schedule. There are buildings, room and lectures
+happening in these rooms. The schema is a fairly straightforward one:
+
+```python
+class Room(models.Model):
+    building = models.CharField(max_length=64, null=False, blank=False)
+    number = models.IntegerField(null=False, blank=False)
+
+class Lecture(models.Model):
+    title = models.CharField(max_length=120, null=False, blank=False)
+    room = models.ForeignKey(Room, null=False)
+    date = models.DateTimeField(null=False, blank=False)
+```
+
+The most obvious case where the ORM falls short, is any non-trivial
+aggregation. Let's say we would like to know which months are the most (or
+least) busy, per building, so we can plan construction work on the campus.
+
+With SQL, that's easy. First, fetch number of lectures happening in each
```

Comment:
I was not even aware that there are such things like window functions or common table expressions (learned SQL years ago) so the question is if these are supported by all databases? Quick search (might be not accurate) says that MySQL doesn't support window functions. I think it's worth mentioning that support for these might vary as it seems to be a reason that ORM's are not supporting such things well.